packages/qwik/src/core/util/string.ts

```diff
@@ -0,0 +1,149 @@
+//   (two-four bytes utf-8)
+// BOM: esc + 0x08fe (four bytes utf-8)
+//
+// If the length of the bytes is odd, the last byte XX is put before the escape
+// character as 0x00XX ESC.
+//
+const ESC = 0x007f;
+const SURROGATE_OFFSET = 0xd800 - 0x0040;
+export const packUint8Array = (bytes: Uint8Array) => {
+  const odd = bytes.length % 2 === 1;
+  const length = bytes.length - (odd ? 1 : 0);
+  let code = '';
+  let surrogate = false;
+  for (let i = 0; i < length; i += 2) {
+    const c = (bytes[i + 1] << 8) | bytes[i];
+    // test high surrogate
+    if (c >= 0xd800 && c <= 0xdbff) {
+      if (surrogate) {
+        // unmatched high surrogate
+        // omit the MSB to make it a normal codepoint
+        const prev = (bytes[i - 1] << 8) | bytes[i - 2];
+        code += String.fromCharCode(ESC, prev - SURROGATE_OFFSET);
+      }
+      surrogate = true;
+      continue;
+    }
+    // test low surrogate
+    if (c >= 0xdc00 && c <= 0xdfff) {
+      if (surrogate) {
+        // valid surrogate pair
+        const prev = (bytes[i - 1] << 8) | bytes[i - 2];
+        code += String.fromCharCode(prev, c);
+        surrogate = false;
+        continue;
+      }
+      // unmatched low surrogate
+      // omit the MSB to make it a normal codepoint
+      const x = c - SURROGATE_OFFSET;
+      code += String.fromCharCode(ESC, x === ESC ? 0x08fd : x);
+      continue;
+    }
+    if (surrogate) {
+      // no low surrogate after high surrogate
+      // omit the MSB to make it a normal codepoint
+      const prev = (bytes[i - 1] << 8) | bytes[i - 2];
+      const x = prev - SURROGATE_OFFSET;
+      code += String.fromCharCode(ESC, x === ESC ? 0x08fd : x);
+      surrogate = false; // reset surrogate
+    }
+    // escape the BOM
+    if (c === 0xfeff) {
+      // BOM
+      code += String.fromCharCode(ESC, 0x08fe);
+      continue;
+    }
+    // extra comaction againt the strinfigy of the control characters
+    if (c <= 0x001f) {
+      code += String.fromCharCode(ESC, c + 0x0020);
+      continue;
+    }
+    // double the escape character
+    if (c === ESC) {
+      code += String.fromCharCode(0x08ff);
+    }
+    // normal codepoint
+    code += String.fromCharCode(c);
+  }
+  if (surrogate) {
+    // ended with unmatched high surrogate
+    // subtract the offset to make it a normal codepoint
+    const c = (bytes[length - 1] << 8) | bytes[length - 2];
+    const x = c - SURROGATE_OFFSET;
+    code += String.fromCharCode(ESC, x === ESC ? 0x08fd : x);
+  }
+  if (odd) {
+    // put the last byte
+    code += String.fromCharCode(bytes[bytes.length - 1], ESC);
+  }
+  return code;
+};
+
+// unpack encoded valid UTF-16 string into Uint8Array
+export const unpackUint8Array = (code: string) => {
+  const odd = code.charCodeAt(code.length - 1) === ESC;
+  const bytes = new Uint8Array(code.length * 2);
+  let j = 0;
+  for (let i = 0; i < code.length; i++) {
+    const c = code.charCodeAt(i);
```

Comment:
change this to `for (const s of code) { const c = s.charCodeAt(0);`, it's [vastly faster](https://jsben.ch/Ng3YQ).

That does mean that we now need to maintain the `escaped` state, which actually makes the code easier to understand.